File: apps/desktop/src/github-feedback/issueFormUrl.test.ts
import assert from 'node:assert/strict'
import { test } from 'node:test'
import {
  buildIssueFormUrl,
  crashTitle,
  isPlaceholderRepo,
  MAX_QUERY_CHARS,
} from './issueFormUrl.ts'

test('crashTitle formats fingerprint and type', () => {
  assert.equal(crashTitle('A1B2C3D4E5F6ffff', 'TypeError: x'), '[crash] a1b2c3d4e5f6 TypeError')
})

test('placeholder repos get an empty issue URL', () => {
  assert.equal(isPlaceholderRepo(''), true)
  assert.equal(isPlaceholderRepo('OWNER/REPO'), true)
  assert.equal(buildIssueFormUrl('OWNER/REPO', 'crash_report.yml', { description: 'x' }).url, '')
})

test('prefills small fields', () => {
  const built = buildIssueFormUrl('acme/app', 'crash_report.yml', {
    title: '[crash] abc TypeError',
    description: 'boom',
  })
  assert.equal(built.bodyTooLarge, false)
  assert.equal(built.url.includes('github.com/acme/app/issues/new'), true)
  assert.equal(built.url.includes('template=crash_report.yml'), true)
})

test('drops a large stack from the query string', () => {
  const stack = 'x'.repeat(MAX_QUERY_CHARS + 500)
  const built = buildIssueFormUrl('acme/app', 'crash_report.yml', { stack, title: '[crash] ab' })
  assert.equal(built.bodyTooLarge, true)
  assert.equal(built.url.length < MAX_QUERY_CHARS, true)
  assert.equal(built.url.includes(stack.slice(0, 40)), false)
  assert.equal(built.clipboardMarkdown, stack)
})

File: examples/android/app/src/test/java/dev/foss/goldenpath/githubfeedback/IssueFormUrlTest.kt
package dev.foss.goldenpath.githubfeedback

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class IssueFormUrlTest {
    @Test
    fun placeholderRepoIsEmpty() {
        assertEquals("", IssueFormUrl.build("OWNER/REPO", "crash_report.yml", mapOf("description" to "x")).url)
    }

    @Test
    fun crashTitleFormats() {
        assertEquals("[crash] a1b2c3d4e5f6 TypeError", IssueFormUrl.crashTitle("A1B2C3D4E5F6ffff", "TypeError: x"))
    }

    @Test
    fun largeStackUsesClipboard() {
        val stack = "x".repeat(IssueFormUrl.MAX_QUERY_CHARS + 500)
        val built = IssueFormUrl.build("acme/app", "crash_report.yml", mapOf("stack" to stack, "title" to "[crash] ab"))
        assertTrue(built.bodyTooLarge)
        assertTrue(built.url.length < IssueFormUrl.MAX_QUERY_CHARS)
        assertFalse(built.url.contains(stack.take(40)))
        assertEquals(stack, built.clipboardMarkdown)
    }
}
